flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -145,6 +146,9 @@
 	/** All registered listeners for status updates of the ResourceManager. */
 	private ConcurrentMap<String, InfoMessageListenerRpcGateway> infoMessageListeners;
 
+	/** The number of failed containers since the master became active. */
+	protected AtomicInteger failedContainerSoFar = new AtomicInteger(0);
```

Comment:
Do we Atomic here? Is there concurrent access?

Reply:
There are three code paths that will create failed container. One is the container can't be started from YARN, which is notified by NMClient callback, the other two are failed to registerTastManager and disconnectTaskManager (When tasks fail). Thus, the concurrent is between akka thread and main thread?